File: src/main/scala/org/tindalos/principle/domain/agents/structure/PackageCohesionModule.scala
package org.tindalos.principle.domain.agents.structure

import org.tindalos.principle.domain.agents.structure.Graph.{NodeId, Node}
import org.tindalos.principle.domain.agents.structure.Structure.{NodeGroup}

import scala.annotation.tailrec

import scala.collection.immutable.Seq

object PackageCohesionModule {

  type PackageName = String

  case class CPackage(name: PackageName, component: NodeGroup)

  def packageOf(nodeId:NodeId) = nodeId.substring(0, nodeId.lastIndexOf('.'))

  def nodesInPackage(nodes:Set[Node], packageName: PackageName) =
    nodes.filter(n ⇒ PackageCohesionModule.packageOf(n.id).equals(packageName))

  val groupByPackages = (rootPackage: PackageName, nodes: Set[Node]) ⇒ {
    //packages with direct classes
    val endPackageNames = nodes.map(n ⇒ packageOf(n.id))
    val packageNames = endPackageNames.flatMap(p ⇒ getPackageNames(rootPackage, p))
    def findNodesInPackageRecursively(packageName: PackageName) = nodes.filter(n ⇒ n.id.startsWith(packageName))
    def buildPackage(pn: PackageName) = (pn, findNodesInPackageRecursively(pn))
    packageNames.map(buildPackage)
  }

  val componentsFromPackages = (rootPackage: PackageName, ns: Set[Node]) ⇒
    groupByPackages(rootPackage, ns)
      .map(x ⇒ (x._1, NodeGroup(x._2)))

  def packageCohesions(rootPackage: PackageName, ns: Set[Node]): Seq[(PackageName, NodeGroup, Double)] =
    componentsFromPackages(rootPackage, ns)
      .filter(x ⇒ x._2.nodes.size > 1)
      .map(x ⇒ (x._1, x._2, x._2.cohesion()))
      .toList.sortBy(_._3).reverse

  //returns all the package names under ...
  def getPackageNames(rootPackage: PackageName, p: PackageName):Set[PackageName] = {
    @tailrec
    def f(acc: Set[PackageName], pn: PackageName): Set[PackageName] =
      if (rootPackage.equals(pn)) acc
      else f(acc + pn, pn.substring(0, pn.lastIndexOf('.')))
    f(Set(),p)
  }
}

File: src/main/java/org/tindalos/principle/domain/analyzers/structure/PackageCohesionModule.java
package org.tindalos.principle.domain.analyzers.structure;

import org.tindalos.principle.domain.core.Node;

import java.util.HashSet;
import java.util.Map;
import java.util.Set;
import java.util.stream.Collectors;

/**
 * Utility operations for grouping classes by package and deriving package cohesion inputs.
 */
public final class PackageCohesionModule {

    private PackageCohesionModule() {
    }

    public static String packageOf(String nodeId) {
        return nodeId.substring(0, nodeId.lastIndexOf('.'));
    }

    public static Set<Node> nodesInPackage(Set<Node> nodes, String packageName) {
        return nodes.stream()
                .filter(node -> packageOf(node.id()).equals(packageName))
                .collect(Collectors.toUnmodifiableSet());
    }

    public static Set<Map.Entry<String, Set<Node>>> groupByPackages(String rootPackage, Set<Node> nodes) {
        var endPackageNames = nodes.stream()
                .map(node -> packageOf(node.id()))
                .collect(Collectors.toSet());

        var packageNames = endPackageNames.stream()
                .flatMap(packageName -> getPackageNames(rootPackage, packageName).stream())
                .collect(Collectors.toSet());

        return packageNames.stream()
                .map(packageName -> Map.entry(packageName, findNodesInPackageRecursively(packageName, nodes)))
                .collect(Collectors.toUnmodifiableSet());
    }

    public static Set<Map.Entry<String, NodeGroup>> componentsFromPackages(String rootPackage, Set<Node> nodes) {
        return groupByPackages(rootPackage, nodes).stream()
                .map(entry -> Map.entry(entry.getKey(), new NodeGroup(entry.getValue())))
                .collect(Collectors.toUnmodifiableSet());
    }

    public static java.util.List<PackageCohesion> packageCohesions(String rootPackage, Set<Node> nodes) {
        return componentsFromPackages(rootPackage, nodes).stream()
                .filter(entry -> entry.getValue().nodes().size() > 1)
                .map(entry -> new PackageCohesion(entry.getKey(), entry.getValue(), entry.getValue().cohesion()))
                .sorted(java.util.Comparator.comparingDouble(PackageCohesion::cohesion).reversed())
                .toList();
    }

    public static Set<String> getPackageNames(String rootPackage, String packageName) {
        var result = new HashSet<String>();
        var current = packageName;

        while (!rootPackage.equals(current)) {
            result.add(current);
            current = current.substring(0, current.lastIndexOf('.'));
        }

        return Set.copyOf(result);
    }

    private static Set<Node> findNodesInPackageRecursively(String packageName, Set<Node> nodes) {
        return nodes.stream()
                .filter(node -> node.id().startsWith(packageName))
                .collect(Collectors.toUnmodifiableSet());
    }

    public record PackageCohesion(String packageName, NodeGroup component, double cohesion) {
    }
}